### PR TITLE
[Azure] Wait Azure resource group to be deleted instead of error out

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes the issue in #3700 where a newly deleted resource group cannot be provisioned immediately:

To reproduce:
`sky launch -c test-azure --cloud azure --cpus 2 echo hi;  sky down test-azure; sky launch -c test-azure --cloud azure --cpus 2 echo hi`

The above script will error out, while this PR fixes it:
```
I 07-01 22:34:20 cloud_vm_ray_backend.py:737] ====== stderr ======
2024-07-01 22:34:18,444 WARNING util.py:259 -- Dropping the empty legacy field head_node. head_nodeis not supported for ray>=2.0.0. It is recommended to removehead_node from the cluster config.
2024-07-01 22:34:18,444 WARNING util.py:259 -- Dropping the empty legacy field worker_nodes. worker_nodesis not supported for ray>=2.0.0. It is recommended to removeworker_nodes from the cluster config.
2024-07-01 22:34:18,444 INFO util.py:382 -- setting max workers for head node type to 0
Traceback (most recent call last):
  File "/var/tmp/skypilot_ray_up__wwue_40.py", line 77, in <module>
    sdk.create_or_update_cluster('/home/gcpuser/.sky/generated/test-azure.yml', **{'no_restart': True})
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/sdk/sdk.py", line 38, in create_or_update_cluster
    return commands.create_or_update_cluster(
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/_private/commands.py", line 314, in create_or_update_cluster
    config = _bootstrap_config(config, no_config_cache=no_config_cache)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/ray/autoscaler/_private/commands.py", line 416, in _bootstrap_config
    resolved_config = provider_cls.bootstrap_config(config)
  File "/home/gcpuser/skypilot/sky/skylet/providers/azure/node_provider.py", line 433, in bootstrap_config
    return bootstrap_azure(cluster_config)
  File "/home/gcpuser/skypilot/sky/skylet/providers/azure/config.py", line 47, in bootstrap_azure
    config = _configure_resource_group(config)
  File "/home/gcpuser/skypilot/sky/provision/common.py", line 285, in wrapper
    return func(*args, **kwargs)
  File "/home/gcpuser/skypilot/sky/skylet/providers/azure/config.py", line 83, in _configure_resource_group
    rg_create_or_update(resource_group_name=resource_group, parameters=params)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/azure/core/tracing/decorator.py", line 94, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/azure/mgmt/resource/resources/v2022_09_01/operations/_operations.py", line 10596, in create_or_update
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/opt/conda/envs/sky/lib/python3.10/site-packages/azure/core/exceptions.py", line 161, in map_error
    raise error
azure.core.exceptions.ResourceExistsError: (ResourceGroupBeingDeleted) The resource group 'test-azure-a374-eastus' is in deprovisioning state and cannot perform this operation.
Code: ResourceGroupBeingDeleted
Message: The resource group 'test-azure-a374-eastus' is in deprovisioning state and cannot perform this operation.
```

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/skylet/providers/azure/config.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 UNIQUE_ID_LEN = 4
 _WAIT_NSG_CREATION_NUM_TIMEOUT_SECONDS = 600
+_WAIT_FOR_RESOURCE_GROUP_DELETION_TIMEOUT = 480  # 8 minutes
```
Comment (cblmemo) [inline sky/skylet/providers/azure/config.py:21]:
```suggestion
_WAIT_FOR_RESOURCE_GROUP_DELETION_TIMEOUT_SECONDS = 480  # 8 minutes
```
nit

File: sky/skylet/providers/azure/config.py [hunk 2]
```diff
@@ -80,7 +82,27 @@ def _configure_resource_group(config):
...
+        except azure.exceptions().ResourceExistsError as e:
+            if "ResourceGroupBeingDeleted" in str(e):
+                if retry % 5 == 0:
+                    # TODO(zhwu): This should be shown in terminal for better
+                    # UX, which will be achieved after we move Azure to use
+                    # SkyPilot provisioner.
+                    logger.warning(
+                        f"Azure resource group {resource_group} is being deleted. "
+                        "It can only be provisioned after it is fully deleted. "
+                        "Waiting..."
```
Comment (cblmemo) [inline sky/skylet/providers/azure/config.py:100]: Should we add some info about why this happened? sth like `it is due to another cluster with the same cluster name being deleted`

**Review by cblmemo — APPROVED** [review]
Thanks for the quick fix! LGTM 🫡

File: sky/skylet/providers/azure/config.py (hunk 1 above)
Comment (Michaelvll) [reply sky/skylet/providers/azure/config.py]: Good point! Fixed. Thanks!

**Pull request merged**